src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,215 @@
+            {
+                var dirPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                Directory.CreateDirectory(dirPath);
+                await DownloadFileContainerAsync(projectId, buildArtifact, dirPath, minimatchFilters, cancellationToken);
+            }
+        }
+
+        private (long, string) ParseContainerId(string resourceData)
+        {
+            // Example of resourceData: "#/7029766/artifacttool-alpine-x64-Debug"
+            var segments = resourceData.Split('/');
+            long containerId;
+            if (segments.Length < 3)
+            {
+                throw new ArgumentException($"Resource data value '{resourceData}' is invalid.");
+            }
+            if (segments.Length >= 3 && segments[0] == "#" && long.TryParse(segments[1], out containerId))
+            {
+                var artifactName = String.Join('/', segments[2]);
+                return(
+                        containerId,
+                        artifactName
+                    );
+            }
+            else
+            {
+                var message = $"Resource data value '{resourceData}' is invalid.";
+                throw new ArgumentException(message, "resourceData");
+            }
+        }
+
+        private async Task DownloadFileContainerAsync(Guid projectId, BuildArtifact artifact, string rootPath, IEnumerable<string> minimatchPatterns, CancellationToken cancellationToken)
+        {
+            var containerIdAndRoot = ParseContainerId(artifact.Resource.Data);
+            
+            var items = await containerClient.QueryContainerItemsAsync(
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            tracer.Info($"Start downloading FCS artifact- {artifact.Name}");
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, tracer);
+            if (minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs, containerIdAndRoot.Item2);
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
+            Parallel.ForEach(folderItems, (folder) =>
+            {
+                var targetPath = ResolveTargetPath(rootPath, folder, containerIdAndRoot.Item2);
+                Directory.CreateDirectory(targetPath);
+            });
+
+            var fileItems = items.Where(i => i.ItemType == ContainerItemType.File);
+
+            var batchItemsBlock = new BatchBlock<FileContainerItem>(
+                batchSize: 8,
+                new GroupingDataflowBlockOptions()
+                {
+                    CancellationToken = cancellationToken,
+                });
+
+            var fetchStream = NonSwallowingTransformManyBlock.Create<IEnumerable<FileContainerItem>, (Func<Task<Stream>>, string)>(
```

Comment:
I think we can remove this block. It doesn't provide much value now. Just merge its logic into the download block